File: javascript/128_Longest_Consecutive_Sequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function(nums) {
    const numSet = new Set(nums);
    let longest = 0;

    for (n of numSet) {
        if (!(numSet.has(n-1))) {
            let length = 0;

            while (numSet.has(n + length)) length++;

            longest = Math.max(length, longest);
        }
    }
    return longest;
};

File: python/128_Longest_Consecutive_Sequence.py
'''
set(): O(1) lookup/insert/delete
Time: O(n)
Space: O(n)
'''

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        numSet = set(nums)
        longest = 0

        for n in numSet:
            if n-1 not in numSet:  # if n is the first number of a sequence
                length = 0

                while n+length in numSet:  # first number: n + 0 == n
                    length += 1
                
            longest = max(longest, length)

        return longest